stlab/concurrency/router.hpp

```diff
@@ -0,0 +1,187 @@
+/*
+Copyright 2015 Adobe
+Distributed under the Boost Software License, Version 1.0.
+(See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+*/
+
+/**************************************************************************************************/
+
+#ifndef STLAB_CONCURRENCY_ROUTER_HPP
+#define STLAB_CONCURRENCY_ROUTER_HPP
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+#include <stlab/concurrency/channel.hpp>
+
+/**************************************************************************************************/
+
+namespace stlab {
+
+/**************************************************************************************************/
+
+inline namespace v1 {
+/**************************************************************************************************/
+
+template <typename T>
+using channel_t = std::pair<sender<T>, receiver<T>>;
+
+namespace detail {
+
+template <typename T, typename K, typename E, typename F>
+class shared_router : public std::enable_shared_from_this<shared_router<T, K, E, F>> {
+    using route_pair_t = std::pair<K, channel_t<T>>;
+    E _executor; // of the router function
+    F _router_function;
+    std::vector<route_pair_t> _routes;
+    bool _ready = false;
+
+public:
+    shared_router() = default;
+    ~shared_router() = default;
+    shared_router(shared_router&&) = default;
+    shared_router(const shared_router&) = default;
+    shared_router& operator=(shared_router&&) = default;
+    shared_router& operator=(const shared_router&) = default;
+
+    shared_router(E executor, F router_function) :
+        _executor{executor}, _router_function(std::move(router_function)) {}
+
+    template <typename... U>
+    shared_router(E executor, F router_function, std::pair<K, channel_t<U>>... route_pairs) :
+        _executor{std::move(executor)},
+        _router_function(std::move(router_function)), _routes{
+                                                          std::make_pair(route_pairs.first,
+                                                                         route_pairs.second)...} {
+        assert(std::is_sorted(std::cbegin(_routes), std::cend(_routes),
+                              [](const auto& x, const auto& y) { return x.first < y.first; }));
+        set_ready();
```

Comment:
Should we really force the call to `set_ready()` here? To my knowledge no other library objects do this. More importantly if you call `set_ready()` in the constructor then you are unable to use `route(key)` to setup the routes after the router is created. You would be forced to setup the routes first then create the router. Look at my example usage to see what I mean.